### PR TITLE
Add tests for Fits scripts

### PR DESCRIPTION
**Pull request by saimn**
Based on #5858 for `fitsheader` (and thanks to @mohanagr). Address #5799

### CODE REVIEW

**Comment by pllim** [comment]
Error seems real:
```
>           raise AssertionError('\n'.join(msg))
E           AssertionError: File(s) not closed:
E             /tmp/fits-test-chboyfm9/diff.txt
```

**Comment by saimn** [comment]
Hmm yes, it's because the output file option uses a logging `FileHandler` to get the messages. It should be ok now.

**Comment by bsipocz** [comment]
Thanks for finishing this up, I think it's really important to get the command line tools tested. 
However I'm not sure it's worth picking up commits from abandoned gsoc PRs when hardly a line would remain from them at the end.

**Comment by MSeifert04** [comment]
Well, some parts were used even if others were completely rewritten so I think it's kind of "fair" to keep it. Either way this LGTM.

File: astropy/io/fits/scripts/fitsdiff.py [hunk 1]
```diff
@@ -292,7 +292,7 @@ def main():
     if opts.quiet:
         out_file = None
     elif opts.output_file:
-        out_file = open(opts.output_file, 'wb')
+        out_file = open(opts.output_file, 'w')
```
Comment (MSeifert04) [inline astropy/io/fits/scripts/fitsdiff.py]: Any specific reason for this change?

File: astropy/io/fits/scripts/fitsdiff.py [hunk 2]
```diff
@@ -319,3 +319,7 @@ def main():
     finally:
         if close_file:
             out_file.close()
+        for handler in log.handlers:
```
Comment (MSeifert04) [inline astropy/io/fits/scripts/fitsdiff.py]: Could you include a comment why this is necessary?

File: astropy/io/fits/tests/test_fitsdiff.py [hunk 3]
```diff
@@ -57,9 +45,73 @@ def test_onediff(self):
...
+        with open(self.temp('diff.txt')) as f:
+            out = f.read()
+        assert out.splitlines()[-4:] == [
+            '     Data differs at [1, 2]:',
+            '        a> 10',
+            '        b> 12',
+            '     1 different pixels found (1.00% different).']
+
+    def test_atol(self):
+        a = np.arange(100, dtype=np.float).reshape((10, 10))
```
Comment (MSeifert04) [inline astropy/io/fits/tests/test_fitsdiff.py]: You don't need the `()` inside the reshape `reshape(10, 10)` should also work

File: astropy/io/fits/scripts/fitsdiff.py (hunk 1 above)
Comment (saimn) [reply astropy/io/fits/scripts/fitsdiff.py]: Yes, it was not working :wink: (writing unicode strings to a file in binary mode). 
Actually I tried to remove this as `diff.report` supports writing to a filename, but the same file is also used to capture the logging output and this was causing conflicts. All this is clearly not optimal, but the goal here is to add tests, I refrained myself to go further!

File: astropy/io/fits/scripts/fitsdiff.py (hunk 2 above)
Comment (saimn) [reply astropy/io/fits/scripts/fitsdiff.py]: Yep.

File: astropy/io/fits/tests/test_fitsdiff.py (hunk 3 above)
Comment (saimn) [reply astropy/io/fits/tests/test_fitsdiff.py]: Sure. While you were here you could also note the `np.float` :stuck_out_tongue:

Comment (MSeifert04) [reply astropy/io/fits/tests/test_fitsdiff.py]: oh, did I forget to submit that comment? 😓

File: astropy/io/fits/tests/test_fitsdiff.py [hunk 4]
```diff
@@ -57,9 +45,73 @@ def test_onediff(self):
...
         hdu_a.writeto(tmp_a)
         hdu_b.writeto(tmp_b)
-        testargs = [tmp_a, tmp_b]
-        sys.argv += testargs
-        numdiff = fitsdiff.main()
+        numdiff = fitsdiff.main([tmp_a, tmp_b])
+        assert numdiff == 1
+
+    def test_manydiff(self, capsys):
+        a = np.arange(100).reshape((10, 10))
```
Comment (MSeifert04) [inline astropy/io/fits/tests/test_fitsdiff.py]: See my comment above regarding `reshape`.

File: astropy/io/fits/scripts/fitsdiff.py (hunk 1 above)
Comment (MSeifert04) [reply astropy/io/fits/scripts/fitsdiff.py]: okay, that makes sense. Weird design there ...

**Comment by MSeifert04** [comment]
Seems like CircleCI hasn't picked up the latest commit... Could you re-commit?

**Comment by saimn** [comment]
@MSeifert04 - I restarted the build

**Comment by pllim** [comment]
@saimn, do you plan to address https://github.com/astropy/astropy/pull/6621#issuecomment-332806505 ?

**Comment by bsipocz** [comment]
@pllim - No need to address that now, it was more like a philosophical comment for the future.

**Comment by pllim** [comment]
OK merging then since it is approved. Thanks!

**Comment by saimn** [comment]
@pllim @bsipocz - I could have done without the PRs, not sure what solution was the best.

**Comment by pllim** [comment]
@saimn , it is merged, so don't sweat it. 😉 Thanks for wrapping this up!

**Pull request merged**